libmamba/src/core/satisfiability_error.cpp

```diff
@@ -0,0 +1,331 @@
+            }
+
+        private:
+            MSolver const& m_solver;
+            MPool const& m_pool;
+            graph_t m_graph;
+            conflict_map m_conflicts;
+            std::map<SolvId, node_id> m_solv2node;
+            node_id m_root_node;
+
+            /**
+             * Add a node and return the node id.
+             *
+             * If the node is already present and ``update`` is false then the current
+             * node is left as it is, otherwise the new value is inserted.
+             */
+            auto add_solvable(SolvId solv_id, node_t&& pkg_info, bool update = true) -> node_id;
+
+            auto add_conflict(node_id n1, node_id n2) -> void;
+            [[nodiscard]] auto add_expanded_deps_edges(node_id from_id,
+                                                       SolvId dep_id,
+                                                       edge_t const& edge) -> bool;
+
+            auto parse_problems() -> void;
+        };
+
+        auto ProblemsGraphCreator::add_solvable(SolvId solv_id, node_t&& node, bool update)
+            -> node_id
+        {
+            if (auto const iter = m_solv2node.find(solv_id); iter != m_solv2node.end())
+            {
+                auto const id = iter->second;
+                if (update)
+                {
+                    m_graph.node(id) = std::move(node);
+                }
+                return id;
+            }
+            auto const id = m_graph.add_node(std::move(node));
+            m_solv2node[solv_id] = id;
+            return id;
+        };
+
+        auto ProblemsGraphCreator::add_conflict(node_id n1, node_id n2) -> void
+        {
+            m_conflicts[n1].insert(n2);
+            m_conflicts[n2].insert(n1);
+        }
+
+        auto ProblemsGraphCreator::add_expanded_deps_edges(node_id from_id,
+                                                           SolvId dep_id,
+                                                           edge_t const& edge) -> bool
+        {
+            auto added = false;
+            for (const auto& solv_id : m_pool.select_solvables(dep_id))
+            {
+                added = true;
+                auto pkg_info = PackageInfo(pool_id2solvable(m_pool, solv_id));
+                auto to_id = add_solvable(
+                    solv_id, PackageNode{ std::move(pkg_info), std::nullopt }, false);
+                m_graph.add_edge(from_id, to_id, edge);
+            }
+            return added;
+        }
+
+        auto ProblemsGraphCreator::parse_problems() -> void
+        {
+            for (const auto& problem : m_solver.all_problems_structured())
+            {
+                // These functions do not return a reference so we make sure to
+                // compute the value only once.
+                auto source = problem.source();
+                auto target = problem.target();
+                auto dep = problem.dep();
```

Comment:
Types should be used instead of auto. If the type is complicate, consider creating aliases.